### PR TITLE
Jobs dashboard

### PR DESCRIPTION
**Pull request by zacwood9**
Jobs share lots in common. By default, in order to provide an interface for viewing and running jobs, IHP users would have to create separate controllers, views, and actions for each job type individually  which is error prone and just not fun. The goal of this job dashboard to use leverage Haskell's type programming and the similar type characteristics of jobs to produce a customizable dashboard for interacting with an IHP application's jobs.

In the future I could imagine this being extended to allow for scheduling of jobs and other nice features.

See included documentation for details https://github.com/digitallyinduced/ihp/pull/832/files#diff-b046ff406e82435d7978bf6566eab8772b94e3b8092a3d8c389d173a0ebfbe9dR94

TODO
- [x] Limit views shown on index page
- [x] Add paginated view page for each job type

### CODE REVIEW

**Comment by mpscholten** [comment]
Just a quick update: Will do a first review + testing tomorrow :) Thanks already for this amazing feature 💯

**Comment by zacwood9** [comment]
Great, thanks :D I've just added some documentation to the Jobs page on the Guide, so refer to that for a explanation of how to use!

File: IHP/Job/Dashboard.hs
Comment (mpscholten) [inline IHP/Job/Dashboard.hs:8]: Just an idea: [we could a code owner file](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners), then github could automatically request e.g. code reviews from you when someone touches the `Dashboard.hs` file :)

File: Guide/jobs.markdown
Comment (mpscholten) [inline Guide/jobs.markdown:106]: Could we make this zero-config out of the box? E.g. we could detect all job queue tables automatically by looking for tables with the `_jobs` suffix :)

File: IHP/Job/Dashboard.hs
Comment (mpscholten) [inline IHP/Job/Dashboard.hs:477]: Could we apply a `timeAgo` here? :) Make it more human readable

File: IHP/Job/Dashboard.hs
Comment (mpscholten) [inline IHP/Job/Dashboard.hs:336]: To simplify the interface we could remove the `Job type:` here (the `title` already has `jobs` in it, so this is kind of redundant)

What do you think about applying a table-name-to-human-readable-name transformation (e.g. like the forms is doing here https://github.com/digitallyinduced/ihp/blob/master/IHP/View/Form.hs#L738)?

So e.g. 

image


would become this:

image

File: IHP/Job/Dashboard.hs
Comment (mpscholten) [inline IHP/Job/Dashboard.hs:481]: in our ihp cloud dashboards we e.g. have a popover that shows the last error if you hover over a failed job:

image

Could we have this here as well? :) https://getbootstrap.com/docs/4.6/components/popovers/

File: IHP/Job/Dashboard.hs
Comment (mpscholten) [inline IHP/Job/Dashboard.hs:339]: Could we also apply `table-hover` here? This makes it easier to select the right row when you have many jobs :)

File: IHP/Job/Dashboard.hs
Comment (mpscholten) [inline IHP/Job/Dashboard.hs:344]: Could we also have a retry button inside the table?

image

This could save one click :)

File: IHP/Job/Dashboard.hs
Comment (mpscholten) [inline IHP/Job/Dashboard.hs:520]: Why is this not supported? :) E.g. maybe you want to link to e.g. the `AllJobsAction` from the admin backend

**Comment by mpscholten** [comment]
Good job so far! :) Also thanks for adding the documentation. Just commented a few ideas :)

File: Guide/jobs.markdown
Comment (zacwood9) [reply Guide/jobs.markdown:106]: I don't believe that would be possible with the current design. Since it's all type level I'm not sure how we could dynamically search for table names and get their corresponding types...I'm sure it's possible but nothing is coming to mind atm.

Also, since it gets injected to the user's application and not the IDE, I think adding it to the project by default might have some negative security implications

File: IHP/Job/Dashboard.hs
Comment (zacwood9) [reply IHP/Job/Dashboard.hs:520]: Didn't see a way to do this at first, but I found a way! Just had to restructure some things.

File: Guide/jobs.markdown
Comment (mpscholten) [reply Guide/jobs.markdown:108]: > Since it's all type level I'm not sure how we could dynamically search for table names and get their corresponding types...I'm sure it's possible but nothing is coming to mind atm.

Yes, I think when we want to do this will all tables, we will loose a bit of type safety as we have to check everything at runtime. Basically it would always pass around table names as a string. But this might make the code even a bit simpler.

E.g. the IHP data editor is also using a similiar approach. It's also working with a dynamic table schema which it detects at runtime.

I think this is worth to explore as otherwise you'll always need to manually add your jobs to the dashboard type whenever you add a new job.

> Also, since it gets injected to the user's application and not the IDE, I think adding it to the project by default might have some negative security implications

I'd still keep the required configuration for the authentication. This way it's still safe :)

Comment (zacwood9) [reply Guide/jobs.markdown:108]: > we will loose a bit of type safety as we have to check everything at runtime

I'm not so much worried about the type safety, but rather the loss of functionality that doing this dynamically would come with. By having to specify types, users can (and pretty much _will_, in order for this to be that useful) define custom implementations of the `DisplayableJob`. This allows us to include fields like parent objects (i.e. in Attics, the ScrapeBandJob has a band_id, and its custom implementation fetches the band and displays it in the table).

I found this function [`someSymbolVal`](https://hackage.haskell.org/package/base-4.14.1.0/docs/GHC-TypeLits.html#v:someSymbolVal) which if I'm understanding it correctly allows you to promote a String to a Symbol at runtime. I will look into this, so hopefully we can get the best of both worlds :)

**Comment by zacwood9** [comment]
Okay just pushed a zero-config option! The dashboard type still expects a type list, but it works slightly differently. It first renders everything like it did before using the types provided, but now it dynamically checks all the jobs tables, and if they were not included as types in the list, it renders a default view for them. So when you add a new job, it shows up automatically :) I think this is the best of both worlds.

Example:
```haskell
type AtticsJobDashboard = JobsDashboardController (BasicAuthStatic "zac" "jobs") '[]
```
will render a dashboard for all tables that have the suffix "_jobs". In Attics this is 3 tables

```haskell
type AtticsJobDashboard = JobsDashboardController (BasicAuthStatic "zac" "jobs") '[InitialScrapeJob, NightlyScrapeJob]
```
This renders InitialScrape and NightlyScrape jobs using my custom implementation, and renders FixSongsJobs using the default.

**Comment by mpscholten** [comment]
Looks pretty nice now :) 

Maybe we can also use the user friendly name here (`Build Container Jobs` instead of `build_container_jobs`):

image

Also: can we apply `timeAgo` or `dateTime` to the timestamps?

image

Right now they look very verbose :)

**Comment by mpscholten** [comment]
Any updates?

**Comment by zacwood9** [comment]
I'm working now on designing an API that allows for easy custom job views. I'm trying to make a way for it to be easy to add columns/additional data without having to reimplement all of the paginated table logic. Haven't had too much time to work on this lately but I will try and get this done soon!

**Comment by mpscholten** [comment]
Sounds great 👍  Ping me anytime if you need some input :)

**Comment by zacwood9** [comment]
Quick update: Got everything working as intended. It's turned into quite a bit of code so I'm going to need to spend some time cleaning everything up. I think it will be worth it in the end though :)

File: IHP/Job/Dashboard.hs
Comment (zacwood9) [reply IHP/Job/Dashboard.hs]: Hey @mpscholten could you send me the HTML + CSS you guys use for these? i'd like to get this in before merge but i'm honestly a bit burnt out right now 😅

Comment (mpscholten) [reply IHP/Job/Dashboard.hs]: Yes :)

```haskell
                    renderStatus JobStatusNotStarted = [hsx|<span class="badge badge-secondary">Not Started</span>|]
                    renderStatus JobStatusRunning = [hsx|<span class="badge badge-primary">Running</span>|]
                    renderStatus JobStatusFailed = [hsx|<span class="badge badge-danger" title="Last Error" data-container="body" data-toggle="popover" data-placement="left" data-content={fromMaybe "" (get #lastError job)}>Failed</span>|]
                    renderStatus JobStatusSucceeded = [hsx|<span class="badge badge-success">Succeeded</span>|]
                    renderStatus JobStatusRetry = [hsx|<span class="badge badge-warning" title="Last Error" data-container="body" data-toggle="popover" data-placement="left" data-content={fromMaybe "" (get #lastError job)}>Retry</span>|]
```

```css
.popover-body {
    background-color: #01313f;
    color: rgb(147, 161, 161);
    font-family: Monaco, Menlo, "Ubuntu Mono", Consolas, source-code-pro, monospace;
    font-size: 11px;
}
```

```js
function initPopover() {
    $('[data-toggle="popover"]').popover({ trigger: 'hover click' })
}
$(document).on('ready turbolinks:load', initPopover);
$(initPopover);
```

Comment (zacwood9) [reply IHP/Job/Dashboard.hs]: Thank you!!

**Comment by zacwood9** [comment]
hey @mpscholten this is ready for a final review. Documentation has been updated to include the new APIs

**Review by mpscholten — APPROVED** [review]
Looks good :) Good job 👍

**Comment by mpscholten** [comment]
Thanks, happy to have this merged now :)

**Pull request merged**
